### ISSUE TITLE
InputTransparent logic/intended behavior

### ISSUE THREAD

**Issue by emceelovin** [dotnet/maui#7489]
### Description

InputTransparent logic seems to be incorrect/flawed, and I'd like to know which logic is correct, so that I don't find out that "working" code is actually incorrect in the future. The docs don't mention enough about InputTransparent IMO.

Given this code, the double tap gesture works(as expected?) on Windows, but it does not fall through the Label -> Grid -> ContentView -> ListView on Android, and this has been a thing since like, preview 12 or so. I brought it up in an issue a long time ago, so it's really frustrating:

```xaml
<!-- code for bidding:BidLineView-->
<ContentView xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="OneGreatTeam.Pages.Bidding.BidLineView"
             CascadeInputTransparent="True">
    <Grid RowDefinitions="auto"
          Margin="0, 10, 0, 10"
          ColumnDefinitions="*, *, *, *, *, *">
        <Label Grid.Row="0"
               Grid.Column="0"
               TextColor="White"
               HorizontalTextAlignment="Center"
               VerticalTextAlignment="Center"
               Text="{Binding Line}"/>
        <Label Grid.Row="0"
               Grid.Column="1"
               TextColor="White"
               HorizontalTextAlignment="Center"
               VerticalTextAlignment="Center"
               Text="{Binding Area}"/>
        <Label Grid.Row="0"
               Grid.Column="2"
               TextColor="White"
               HorizontalTextAlignment="Center"
               VerticalTextAlignment="Center"
               Text="{Binding Duration}"/>
        <Label Grid.Row="0"
               Grid.Column="3"
               TextColor="White"
               HorizontalTextAlignment="Center"
               VerticalTextAlignment="Center"
               Text="{Binding RDO}"/>
        <Label Grid.Row="0"
               Grid.Column="4"
               TextColor="White"
               HorizontalTextAlignment="Center"
               VerticalTextAlignment="Center"
               Text="{Binding ALA}"/>
        <Label Grid.Row="0"
               Grid.Column="5"
               TextColor="White"
               HorizontalTextAlignment="Center"
               VerticalTextAlignment="Center"
               Text="{Binding OSM}"/>
    </Grid>
</ContentView>

<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:bidding="clr-namespace:OneGreatTeam.Pages.Bidding"
             x:Class="OneGreatTeam.Pages.Bidding.BidPage"
             Title="{OnIdiom Phone='Bid', Desktop=''}"
             BackgroundColor="#18191A">
    <Grid RowDefinitions="auto, *, auto"
          ColumnDefinitions="*, *, *, *, *, *">
        <Frame CornerRadius="10"
               Padding="0"
               Grid.Row="0"
               Grid.Column="0"
               BackgroundColor="Gray">
            <Label Text="Line"
                   HorizontalTextAlignment="Center"
                   VerticalTextAlignment="Center"/>
        </Frame>
        <Frame CornerRadius="10"
               Padding="0, 5, 0, 5"
               Grid.Row="0"
               Grid.Column="1"
               BackgroundColor="Gray">
            <Label Text="Area"
                   HorizontalTextAlignment="Center"
                   VerticalTextAlignment="Center"/>
        </Frame>
        <Frame CornerRadius="10"
               Padding="0, 5, 0, 5"
               Grid.Row="0"
               Grid.Column="2"
               BackgroundColor="Gray">
            <Label Text="Hours"
                   HorizontalTextAlignment="Center"
                   VerticalTextAlignment="Center"/>
        </Frame>
        <Frame CornerRadius="10"
               Padding="0, 5, 0, 5"
               Grid.Row="0"
               Grid.Column="3"
               BackgroundColor="Gray">
            <Label Text="RDO"
                   HorizontalTextAlignment="Center"
                   VerticalTextAlignment="Center"/>
        </Frame>
        <Frame CornerRadius="10"
               Padding="0, 5, 0, 5"
               Grid.Row="0"
               Grid.Column="4"
               BackgroundColor="Gray">
            <Label Text="ALA"
                   HorizontalTextAlignment="Center"
                   VerticalTextAlignment="Center"/>
        </Frame>
        <Frame CornerRadius="10"
               Padding="0, 5, 0, 5"
               Grid.Row="0"
               Grid.Column="5"
               BackgroundColor="Gray">
            <Label Text="OSM"
                   HorizontalTextAlignment="Center"
                   VerticalTextAlignment="Center"/>
        </Frame>
        <ListView Grid.Row="1"
                  Grid.Column="0"
                  Grid.ColumnSpan="6"
                  SelectionMode="Single"
                  SelectedItem="{Binding SelectedLine}"
                  SeparatorVisibility="None"
                  ItemsSource="{Binding BidLines}">
            <ListView.ItemTemplate>
                <DataTemplate>
                    <ViewCell>
                        <bidding:BidLineView/>
                    </ViewCell>
                </DataTemplate>
            </ListView.ItemTemplate>
           <ListView.GestureRecognizers>
                <TapGestureRecognizer NumberOfTapsRequired="2" Tapped="TapGestureRecognizer_Tapped" Command="{Binding LineSelectedCommand}"/>
           </ListView.GestureRecognizers>
        </ListView>
        <Button Grid.Row="2"
                Grid.Column="0"
                Grid.ColumnSpan="6"
                IsVisible="{OnIdiom Phone=True, Desktop=False}"
                Text="Select Line"/>
    </Grid>
</ContentPage>
```

After reading the docs, what I thought was expected -- and wanted -- behavior on Windows, actually seems 100% incorrect, and Android is the correct behavior. Kind of. Views are not InputTransparent by default(according to MSDN), and children inherit this by default as well, BUT on Windows, when placing a TapGestureRecognizer on the ListView, input is falling through any descendent in the UI tree all the way up to the input gesture registered to it. This is completely against what the documentation mentions. Input should be blocked by any descendent in a tree unless InputTransparent=True. Which causes all the children to inherit this by default due to CascadeInputTransparency=True by default. On Android, input on the ListView is blocked unless you place the TapGestureRecognizer further down the tree and place it on the ```<bidding:BidLineView/>```. IMO, and according to the doc, this is still wrong. Input should be blocked by the Grid layout inside of ```<bidding:BidLineView/>``` and its children on both Android and Windows. Input should be blocked by anything obstructing a UI element in the visual tree. Necessitating the need for InputTransparent/CascadeInputTransparent. Meaning, unless something is the very last element in the tree and/or unobstructred, it should not receive input unless the proper attributes are specified. 

Also, when placing the TapGestureRecognizer on ```<bidding:BidLineView>```, ListView selection doesn't work on the ListView anymore, but the tap gesture event is fired. It's like whack-a-mole. Gain here lose there. 

TapGestureRecognizer on ListView with ```<bidding:BidLineView InputTransparent="True">``` causes Android to crash on a local device and the emulator. Windows is fine.

There really isn't good documentation for input gesture recognition either. How deep does CascadeInputTransparent go? It says all children are affected. Does it end at the layout its specified on, or does it continue until the last child in the layouts visual tree? A layout can have layouts as children, so where does it end? If CascadeInputTransparent is True by default, then I would assume InputTransparent=True would mean it goes all the way until the last element in that node of the tree.

Windows:
ContentPage -> Grid(Tap gesture recognized) -> ListView(Tap gesture recognized) -> bidding:BidLineView(Tap gesture recognized)
Gestures shouldn't be recognized at any of these places.

Android:
ContentPage -> Grid(Nothing) -> ListView(Nothing) -> bidding:BidLineView(Tap gesture recognized/ListView item selection no longer works due to BidLineView intercepting the double tap gesture. Should be blocked by descendants inside BidLineView)
Closer to what the docs mention as behavior, but BidLineView still shouldn't receive any input.

According to the docs and input transparency, the <Label>'s inside of the Grid inside of BidLineView should block everything unless they have gestures registered on them or descendants have InputTransparency="True". Allowing input to bubble upwards.

Can anyone explain the design/blueprint behind input transparency in the UI tree? What are the constraints, what's allowed, how it's SUPPOSE to work, where it's suppose to be placed, examples, etc. Been fighting this problem since the preview stages. All of this also happens with MVVM/Command's. Which is what I actually run with non repro/sample code. Is trying to double tap a ListView item bad design/practice when choosing an item? My workaround, for now, is a Button on Phone/mobile. As Windows/desktop works the way I want it to. Keeping the tap gesture constrained to the main page, versus BidLineView, makes working with the pages ViewModel a lot easier and has proper UI functionality with highlighting and what not.

### Steps to Reproduce

See xaml

### Version with bug

6.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

Android/Windows

### Did you find any workaround?

Yes, had to create a button on Android in lieu of double tap gestures.

**Comment by emceelovin** [comment]
Also received a random exception from what looks like ListView while scrolling/messing with it inside this Page:

[me.onegreattea] runtime.cc:677] JNI DETECTED ERROR IN APPLICATION: JNI CallObjectMethodA called with pending exception java.lang.IllegalStateException: The specified child already has a parent. You must call removeView() on the child's parent first.
[me.onegreattea] runtime.cc:677]   at void android.view.ViewGroup.addViewInner(android.view.View, int, android.view.ViewGroup$LayoutParams, boolean) (ViewGroup.java:6084)
[me.onegreattea] runtime.cc:677]   at void android.view.ViewGroup.addView(android.view.View, int, android.view.ViewGroup$LayoutParams) (ViewGroup.java:5903)
[me.onegreattea] runtime.cc:677]   at void android.view.ViewGroup.addView(android.view.View, int) (ViewGroup.java:5843)
[me.onegreattea] runtime.cc:677]   at void android.view.ViewGroup.addView(android.view.View) (ViewGroup.java:5815)
[me.onegreattea] runtime.cc:677]   at android.view.View crc64e1fb321c08285b90.ListViewAdapter.n_getView(int, android.view.View, android.view.ViewGroup) (ListViewAdapter.java:-2)
[me.onegreattea] runtime.cc:677]   at android.view.View crc64e1fb321c08285b90.ListViewAdapter.getView(int, android.view.View, android.view.ViewGroup) (ListViewAdapter.java:100)
[me.onegreattea] runtime.cc:677]   at android.view.View android.widget.HeaderViewListAdapter.getView(int, android.view.View, android.view.ViewGroup) (HeaderViewListAdapter.java:234)
[me.onegreattea] runtime.cc:677]   at android.view.View android.widget.AbsListView.obtainView(int, boolean[]) (AbsListView.java:2617)
[me.onegreattea] runtime.cc:677]   at android.view.View android.widget.ListView.makeAndAddView(int, int, boolean, int, boolean) (ListView.java:2224)
[me.onegreattea] runtime.cc:677]   at android.view.View android.widget.ListView.fillDown(int, int) (ListView.java:826)
[me.onegreattea] runtime.cc:677]   at void android.widget.ListView.fillGap(boolean) (ListView.java:789)
[me.onegreattea] runtime.cc:677]   at boolean android.widget.AbsListView.trackMotionScroll(int, int) (AbsListView.java:6082)
[me.onegreattea] runtime.cc:677]   at boolean android.widget.ListView.trackMotionScroll(int, int) (ListView.java:2141)
[me.onegreattea] runtime.cc:677]   at void android.widget.AbsListView$FlingRunnable.run() (AbsListView.java:5594)
[me.onegreattea] runtime.cc:677]   at void android.view.Choreographer$CallbackRecord.run(long) (Choreographer.java:1108)
[me.onegreattea] runtime.cc:677]   at void android.view.Choreographer.doCallbacks(int, long, long) (Choreographer.java:866)
[me.onegreattea] runtime.cc:677]   at void android.view.Choreographer.doFrame(long, int, android.view.DisplayEventReceiver$VsyncEventData) (Choreographer.java:792)
[me.onegreattea] runtime.cc:677]   at void android.view.Choreographer$FrameDisplayEventReceiver.run() (Choreographer.java:1092)
[me.onegreattea] runtime.cc:677]   at void android.os.Handler.handleCallback(android.os.Message) (Handler.java:938)
[me.onegreattea] runtime.cc:677]   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:99)
[me.onegreattea] runtime.cc:677]   at boolean android.os.Looper.loopOnce(android.os.Looper, long, int) (Looper.java:226)
[me.onegreattea] runtime.cc:677]   at void android.os.Looper.loop() (Looper.java:313)
[me.onegreattea] runtime.cc:677]   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:8663)
[me.onegreattea] runtime.cc:677]   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
[me.onegreattea] runtime.cc:677]   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:567)
[me.onegreattea] runtime.cc:677]   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:1135)
[me.onegreattea] runtime.cc:677] 
[me.onegreattea] runtime.cc:677]     in call to CallObjectMethodA
[me.onegreattea] runtime.cc:677]     from android.view.View crc64e1fb321c08285b90.ListViewAdapter.n_getView(int, android.view.View, android.view.ViewGroup)
[libc] Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 30360 (me.onegreatteam), pid 30360 (me.onegreatteam)

**Comment by hartez** [comment]
> Also received a random exception from what looks like ListView while scrolling/messing with it inside this Page on Android:
> Was able to reproduce this. When switching to the second tab, and back after selecting a line, then scrolling, this exception is thrown each time. Here is the project: [Repro.zip](https://github.com/dotnet/maui/files/8777012/Repro.zip)

Please open a second issue for your ListView bug so we can track it and address it.

**Comment by emceelovin** [comment]
Copy. Will do!

**Comment by hartez** [comment]
There are two concepts to keep in mind here, and I'll be the first to admit that the documentation is confusing/lacking on both. The first is "input transparency", the second is "gesture bubbling". 

Views have a boolean property `InputTransparent` which defaults to `false`. Views which are input transparent don't respond to touches/clicks - the interaction effectively goes "through" them to whatever is behind them. MAUI Controls also has a Layout property `CascadeInputTransparent`, which is a holdover from Xamarin.Forms. It defaults to `true`, and it just means that if you mark a Layout as `InputTransparent = true`, then every control inside of that Layout will also be effectively input transparent. If you disable `CascadeInputTransparent` and mark a Layout as input transparent, then the Layout itself will not take input, but the controls inside of it can. 

If you don't mark the Layout as `InputTransparent = true`, then the `CascadeInputTransparent` property has no effect. So in your example XAML above, nothing is input transparent. 

The second concept, "gesture bubbling", refers to how gestures defined on parent controls interact with child controls. Say, for example, you have a Layout with a single Label in it. If you define a tap gesture on the Layout, the gesture will fire if you tap on the Label. The label does not have its own gesture, but it bubbles the input up to its parent. This is the behavior for controls which do no accept input of their own (e.g., Labels and Images). Controls which accept input (e.g., Entries and Editors) don't bubble gestures up to their parent controls. 

> How deep does CascadeInputTransparent go? It says all children are affected. Does it end at the layout its specified on, or does it continue until the last child in the layouts visual tree? A layout can have layouts as children, so where does it end? If CascadeInputTransparent is True by default, then I would assume InputTransparent=True would mean it goes all the way until the last element in that node of the tree.

Your assumption is correct - if `CascadeInputTransparent = true` and `InputTransparent = true` on a Layout, then _everything_ contained in that Layout should be input transparent unless explicitly marked otherwise. If that's not the case, then we likely have a bug.

> Is trying to double tap a ListView item bad design/practice when choosing an item?

The underlying control for a ListView on Android is an Android ListView, which has its own selection built it - I wonder if you're seeing a bad interaction with that. It's also entirely possible that we missed something porting the gesture/selection code from Forms and you're seeing a bug. Do your selection problems show up in the Repro.zip project you uploaded in [this comment](https://github.com/dotnet/maui/issues/7489#issuecomment-1138220821)?

**Comment by emceelovin** [comment]
Appreciate the response a ton, and I don't mean to come off harsh at all. I just want to see MAUI succeed so much, and try to offer up as much repro as I can. Will get back when I'm free. Also, I deleted and moved the other issue.

**Comment by emceelovin** [comment]
> Views have a boolean property InputTransparent which defaults to false. Views which are input transparent don't respond to touches/clicks - the interaction effectively goes "through" them to whatever is behind them. MAUI Controls also has a Layout property CascadeInputTransparent, which is a holdover from Xamarin.Forms. It defaults to true, and it just means that if you mark a Layout as InputTransparent = true, then every control inside of that Layout will also be effectively input transparent. If you disable CascadeInputTransparent and mark a Layout as input transparent, then the Layout itself will not take input, but the controls inside of it can.

That is correct, but according to the documentation, the default value of false says an element will receive input. 

> Each visual element has an InputTransparent bindable property that's used to define whether the element can receive input. Its default value is **false,** ensuring that the element **can** receive input.

> If you don't mark the Layout as InputTransparent = true, then the CascadeInputTransparent property has no effect. So in your example XAML above, nothing is input transparent.

This makes sense, and I didn't really realize that. It was pretty late. Since elements default to false, InputTransparent=True is the only value where inheriting it will make a difference.

> The second concept, "gesture bubbling", refers to how gestures defined on parent controls interact with child controls. Say, for example, you have a Layout with a single Label in it. If you define a tap gesture on the Layout, the gesture will fire if you tap on the Label. The label does not have its own gesture, but it bubbles the input up to its parent. This is the behavior for controls which do no accept input of their own (e.g., Labels and Images). Controls which accept input (e.g., Entries and Editors) don't bubble gestures up to their parent controls.

This is what I was talking about. The rules and what not that the documentation do not explain. When and where there will be implicit blocking of input etc. I assumed that an element will block something it obstructs just because that's what it seemed like to me with the flags and what not, so thank you for that. I meant to add that also. Unless I did. I was going to say does the bubbling up of gestures only matter when it comes to points in the tree that have gestures registered. E.g. an Entry like you say, then nothing for lets say two or three elements higher up, then another gesture registered. In which the Entry would then block the other gesture registered so many elements up in the tree. In other words, input transparency only matters when comparing the obstruction of an element with input registered relative to other elements with input registered. I wasn't able to find out any information on that

> The underlying control for a ListView on Android is an Android ListView, which has its own selection built it - I wonder if you're seeing a bad interaction with that. It's also entirely possible that we missed something porting the gesture/selection code from Forms and you're seeing a bug. 

On windows, everything is absolutely perfect. On Android, same code, I haven't been able to register a double tap gesture on a ListView and have it fire with a double tap. I have to place it up or down the tree, from the point of the ListView, for it to work, but then odd behavior starts to happen. Selection is very hit or miss; literally. You have to tap quite a few times to get even the default single tap to register, and then it'll even be behind and start selecting after you stop tapping.

> Do your selection problems show up in the Repro.zip project you uploaded in https://github.com/dotnet/maui/issues/7489#issuecomment-1138220821?

They do unfortunately. If you leave the project the way it is, the double tap will not fire when registered to the ListView itself. On windows it's perfectly fine. If you register the gesture on the Grid, or the <bidding:BidLineView/>, it'll work, but the performance is janky. When registered on the grid, ListView highlighting is EXTREMELY laggy and behind. When registered on bidding:BidLineView, it actually works great. EXCEPT the ListView doesn't receive input so it can highlight the selected item, but it seems like the viewmodel for Pages/BidPage.xaml.cs is being updated as a result of the tap. Just the highlight is missing, so i'm not exactly sure how that's even happening.

**Comment by emceelovin** [comment]
bump

@hartez Were you able to find anything with the repro project?

**Comment by hartez** [comment]
Haven't had a chance to look yet. It's on my list.

**Comment by emceelovin** [comment]
No worries at all bro.

**Comment by mattleibow** [comment]
I wonder if this is fixed now with: https://github.com/dotnet/maui/pull/14650

It was backported to net7.0 SR5

**Comment by mattleibow** [comment]
It may also be related to this #10252

There are several pages and steps in the linked repro. Is it possible to make a simpler one to ease the testing as well as eliminating other factors. I do think this is a real issue, but with the number of pages and several views this will take longer to figure out.

Also, what happens if you swap the ContentView with a Layout? I have seen that ContentView works differently than what may be expected: https://github.com/dotnet/maui/issues/10977#issuecomment-1529303644

**Comment by mattleibow** [comment]
Input transparency should be fixed now with:

 - https://github.com/dotnet/maui/pull/14846

Fixed in [8.0.0-preview.7.8842](https://github.com/dotnet/maui/releases/tag/8.0.0-preview.7.8842)

**Comment by h1nson** [comment]
Simplest workaround: adding an empty TapGestureRecognizer to the view that's over top. Explicitly setting InputTransparent didn't seem to do anything.
